### ISSUE TITLE
New warnings by recent clang++ 3.6.1

### ISSUE THREAD

**Issue by lrineau** [CGAL/cgal#139]
After a discussion on the CGAL-develop mailing list, where I asked developers to try the RC of CGAL-4.6.1...

The following two warnings are scary. I wonder if that can be considered as 
blockers.
```
> ==== Warnings (I probably missed a few; unfortunately, I ====
> ==== did not remember to pipe the full output to a file) ====
> 
> /home/akobel/CGAL-4.6.1/demo/Triangulation_2/../../include/CGAL/Compact_cont
> ainer.h:1150:13: warning:
>        reference cannot be bound to dereferenced null pointer in
> well-defined C++ code;
>        comparison may be assumed to always evaluate to false
>        [-Wtautological-undefined-compare]
>      return &*rhs == NULL;
>              ^~~~    ~~~~
> 
> /home/akobel/CGAL-4.6.1/demo/Triangulation_2/../../include/CGAL/Compact_cont
> ainer.h:1095:15: note:
>        'operator*' returns a reference
>      reference operator*() const { return *(m_ptr.p); }
>                ^
```
Can somebody explain that warnings? That seems an important bug, because 
`it==NULL` is part of the `Handle` concept. Should the code be the following 
instead?
```
  return rhs.operator->() == NULL;
```

```
> /home/akobel/CGAL-4.6.1/examples/Partition_2/../../include/CGAL/vector.h:199
> :15: warning:
>        reference cannot be bound to dereferenced null pointer in
> well-defined C++ code;
>        pointer may be assumed to always convert to true
> [-Wundefined-bool-conversion]
>          if ( &*start_)
>               ~^~~~~~~
> 
> /home/akobel/CGAL-4.6.1/examples/Partition_2/../../include/CGAL/vector.h:75:
> 15: note:
>        'operator*' returns a reference
>      reference operator*()                const { return *ptr; }
>                ^
```
Same here:
```
  if( start_.operator->() )
```
seems a better code.

**Comment by gavinsalam** [comment]
This warning (and a similar one for !=) appears to be a serious one. The piece of code appended below gives the wrong answer when compiled with CGAL-4.6.1 and clang-3.6.1 with optimisation switched on. The expected output on running the code is

    If the following is not '0', there is a bug: 0
    If the following is not '0', there is a bug: 0x0

The actual output is instead

    If the following is not '0', there is a bug: 1
    If the following is not '0', there is a bug: 0x0

With optimisation turned off in clang-3.6.1, or also with g++-4.8 (with or without optimisation) we get the correct answer.

The code is as follows:

```C++
typedef CGAL::Triangulation_vertex_base_with_info_2<int,K> Vbb;
typedef CGAL::Triangulation_hierarchy_vertex_base_2<Vbb> Vb;
typedef CGAL::Triangulation_face_base_2<K> Fb;
typedef CGAL::Triangulation_data_structure_2<Vb,Fb> Tds;
typedef CGAL::Delaunay_triangulation_2<K,Tds>  Dt;
typedef CGAL::Triangulation_hierarchy_2<Dt> Triangulation;

typedef Triangulation::Vertex_handle  Vertex_handle;
typedef Triangulation::Point          Point; /// CGAL Point structure
typedef Triangulation::Vertex_circulator Vertex_circulator;
typedef Triangulation::Face_circulator Face_circulator;
typedef Triangulation::Face_handle Face_handle;

using namespace std;

int main() {
  Vertex_handle v(NULL);
  cout << "If the following is not '0', there is a bug: " << (v != NULL) << endl;
  cout << "If the following is not '0', there is a bug: " << v.operator->() << endl;
}

```

**Comment by lrineau** [comment]
Should be fixed by the merge of #161.

**Comment by gavinsalam** [comment]
Many thanks.
Best regards,
Gavin
